lib/collaborative_filtering.py

```diff
@@ -2,3 +2,212 @@
+        self.partial_train(n_iter)
+        print("sum")
+        print(sum(sum(self.ratings)))
+
+    def partial_train(self, n_iter):
+        """
+        Train model for n_iter iterations. Can be
+        called multiple times for further training.
+        @param n_iter (int) number of iterations
+        """
+        ctr = 1
+        while ctr <= n_iter:
+            if self._v:
+                print('\tcurrent iteration: {}'.format(ctr))
+                print('Error %f' % self.get_mse(self.user_vecs.dot(self.item_vecs.T), self.ratings))
+            self.user_vecs = self.als_step(self.user_vecs,
+                                           self.item_vecs,
+                                           self.ratings,
+                                           self._lambda,
+                                           type='user')
+            self.item_vecs = self.als_step(self.item_vecs,
+                                           self.user_vecs,
+                                           self.ratings,
+                                           self._lambda,
+                                           type='item')
+            ctr += 1
+
+    def get_predictions(self):
+        """
+        Predict ratings for every user and item.
+        @returns (ndarray) predictions
+        """
+        return self.user_vecs.dot(self.item_vecs.T)
+
+    def predict(self, u, i):
+        """
+         Single user and item prediction.
+         @returns float prediction score
+         """
+        return self.user_vecs[u, :].dot(self.item_vecs[i, :].T)
+
+    def calculate_learning_curve(self, iter_array, test):
+        """
+        Keep track of MSE as a function of training iterations.
+        @param (list) iter_array  List of numbers of iterations to train for each step of
+            the learning curve. e.g. [1, 5, 10, 20]
+        @param test (ndarray) Testing dataset (assumed to be user x item).
+        """
+        iter_array.sort()
+        self.train_mse = []
+        self.test_mse = []
+        iter_diff = 0
+        for (i, n_iter) in enumerate(iter_array):
+            if self._v:
+                print('Iteration: {}'.format(n_iter))
+            if i == 0:
+                self.train(n_iter - iter_diff)
+            else:
+                self.partial_train(n_iter - iter_diff)
+            predictions = self.get_predictions()
+            self.train_mse += [self.get_mse(predictions, self.ratings)]
+            self.test_mse += [self.get_mse(predictions, test)]
+            if self._v:
+                print('Train mse: ' + str(self.train_mse[-1]))
+                print('Test mse: ' + str(self.test_mse[-1]))
+            iter_diff = n_ite
+
+    def get_mse(self, pred, actual):
+        """
+        simply calculate the rmse by calling the method on the evaluator
+        @returns (float) root mean squared error
+        """
+        return self.evaluator.get_rmse(pred, actual)
+
+    def get_ratings(self):
```

Comment:
Do you need this function? It seems that you shouldn't need the ratings outside the class, or? unless you need it in the evaluation or so.